Optimization

straw-boy opened the pull request:
Made scaled prox a function of the Family class.

jolars commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ doc
 Meta
 *.pdf
 docs
+*run.r
```
you don't need to change anything, but in the future it is better to avoid entries in gitignore that are specific to you (see <https://stackoverflow.com/questions/1753070/how-do-i-configure-git-to-ignore-some-files-locally>)

jolars commented on src/algorithms/newton_raphson.h:
```diff
@@ -14,29 +14,54 @@ mat Family::newtonRaphson(const T& x, const mat& y, const double rho, const mat&
-    h.diag() += rho;
     
-    mat step = -reshape(solve(h,vectorise(g)),size(z));
+    if (use_woodbury) {
+      // activation = pseudoHessian(y, lin_pred);
+
+      // mat tmp = xxT;
+      // tmp.diag() += rho/activation;
+      // step = x.t()*solve(tmp, x*g) - g;
+      // step /= rho;
+
```
should this really be commented out?

straw-boy replied:
I wanted to ask the same, it produces error due to the badly conditioned matrix. It won't get executed since use_woodbury is false, should I uncomment it?

jolars reviewed (changes requested):
Good work!

Just some feedback on the pull request itself: it is a little bit all over the place. It includes changes that are separate, the title is vague and there is no real description of the changes. I know you might have had to rush this and your previous pull requests have been better, so it's not a big issue.

I think it would be a good idea to spend some time polishing the benchmarks. I suggest log-transforming the y axis and possibly modifying the plot data so that the algorithms that take forever to converge don't destroy the plots. 

I also think you should include benchmarks checking for different levels of regularization. It is much more important to get this right before starting to merge into the SLOPE package since we will not want to include all of the solvers unless there is **strong** evidence that that would be beneficial.

straw-boy commented:
Sorry about that, I was not in the right mental space.
I have added the vignette which does what we discussed over mail. If nothing goes wrong, it should be ready soon. I modified plotting so it plots log of loss in y-axis and plotting is done only until the 2nd fastest solver.  Let's see how this turns out.

straw-boy commented:
Looks like code is breaking for the last benchmark. I'll fix it at night. Meanwhile, do have a look at the new plots. I have increase n for n>p cases to observe if FISTA degraded considerably when compared to PN.
